### ISSUE TITLE
Is is possible to use PDFJs-dist and ngx-extended-pdf-viewer in a same project ?

### ISSUE THREAD

**Issue by latannon** [stephanrauh/ngx-extended-pdf-viewer#799]
Hi,

First of all thank you for this PDF viewer. It is an awesome !

Im a facing some issues to integrate PDFJS-dist and Ngx-extended-pdf-viewer in the same projet. Let's me explain our uses cases just below.
In our projet we render and draw image from PDF page, using pdfjs-dist (2.2.228) and the same way we use ngx-extended-pdf (1.5.1 using a standard PDFJs version 2.2.228). This works like a charm but as you can see versions are very old... 

I am trying to update these packages to latest ones : 

* pdfjs-dist : 2.9.359
* ngx-extended-pdf-viewer : 9.0.1 (internal PDFJs version = 2.9.149 ?)

As I understand since the version 1.8 (I believe) you add a custom version of PDFJs (so a non-standard one in order to add new features like showing unverified visual signatures). We are facing some difficulties with this in our project. Am I right ?

Anyway I do update  packages and now, I am  trying to solve the conflict between PDFJs-dist worker version and the Ngx-extended-pdf-viewer internal PDFjs version.

## If I do not load workerSrc for PDFJs, I can't load a document and get error (obvious ?) : 

`ERROR Error: No "GlobalWorkerOptions.workerSrc" specified.`

```js
PDFJS.getDocument(pdfUrl).promise.then((pdfDoc) => ...));
```
Ngx-extended-pdf-viewer : Works fine !

## If I do specified a workerSrc for PDFJs

```js

 ngOnInit(): void {
    this.loading = true;

    import('pdfjs-dist/build/pdf.worker.entry').then(() => {
        PDFJS.getDocument(pdfUrl).promise.then((pdfDoc) => ...));
        // ...
    });
}
```

* First time I load Ngx-extended-pdf-viewer (wrapper) : it works fine ! 
* PDFJs image rendering component : works fine too.

If I try to load Ngx-extended-pdf-viewer again, it fails with error : 
![image](https://user-images.githubusercontent.com/248874/121237638-91ffa800-c897-11eb-9bd0-d8d3359d1263.png)

And it makes sense ! 

Is there a way to make them cohabitate together in the same project ?

I think I don't understand enough the concept of worker here, so I can't figure out how to address the issue properly...

I hope I was clear enough and you understand my purpose. If you have a hint or tips to achieve my goal it will be nice.

Thank you for reading.

**Comment by stephanrauh** [comment]
You're right, the original pdf.js and my fork have diverged considerably. It's a bad idea to use the Mozilla distribution to power ngx-extended-pdf-viewer. Every once in a while a developers reports to run such a configuration successfully, but in general, it's doomed to fail.

However, I believe it's possible to do it the other way round. Drop the original distribution of pdf.js in favor of my copy. It should do the trick, too. Instead of loading pdf.js from `node_modules/pdfjs-dist/` you can load it from `node_modules/ngx-extended-pdf-viewer/assets`.

If that approach fails, please tell me and I'll fix the bug. More likely than not, it's because, well, I'm lazy and forget to implement a check or two. :)

**Comment by stephanrauh** [comment]
If everything fails, you can also use the brute-force approach I'm using in the showcase to run ng2-pdf-viewer. Instead of using the Angular router I've implemented a navigation with an old-school full-page-reload. That's the only way I know how to unload pdf.js from memory.

**Comment by stephanrauh** [comment]
Another side remark: you've spent a lot of time writing the issue. I appreciate that. Your issue isn't exactly a standard issue. Nonetheless, I was able to understand it quickly. Awesome! Merci beaucoup!

Oh, and please tell me I my hints help you to solve the problem. I prefer to keep issues open until I know for sure they're solved. 

@+
Stephan

**Comment by latannon** [comment]
Hi, 

What reactivity! 

Thank you a lot for taking the time to read and answer that quickly. 

I will try your différent suggestions and take you inform what did work or not.

**Comment by latannon** [comment]
Hi,  thank you again. I try many things and here a small feedback.

## TL;DR:

* Cohabitation between pdfjs-dist and ngx-extended-pdf-viewer : works on modern updated browser, fails on old Firefox 52 version (THE organization official browser 👎) => https://github.com/latannon/pdfjs-and-co/tree/sharing-pdfjs-with-extenteded-pdf-viewer
* Cohabitation between pdfjs-dist and ng2-pdf-viewer: works => https://github.com/latannon/pdfjs-and-co/tree/sharing-pdfjs-with-ng2-pdf-viewer

----------------------
* ngx-extented-pdf-viewer: 

I try to use your internal pdf.worker-es5.js. It work's like a charm (on modern and updated browser like Chromium based ones). In our organization we are stick to the official "administrated" browser : Firefox ESR 52 version (for old applications compatibility) and for understood reason (maybe oldness) it doesn't work there. 
You can see my implementation (POC) here : https://github.com/latannon/pdfjs-and-co/tree/sharing-pdfjs-with-extenteded-pdf-viewer
I try as hard as I understand but fail to make it works!

* ng2-pdf-viewer : 

Since we are not really using extended features (like: zoom, navigation...), I try to use ng2-pdf-viewer and it works (both on Chrome and FF ESR 52)! 
To display unverified signatures (I use piece of code from your version of pdfjs worker) : https://github.com/latannon/pdfjs-and-co/tree/sharing-pdfjs-with-ng2-pdf-viewer

For now as we have to push in production soon, I use ng2-pdf-viewer solution until we can remove FF52 from users computers...

Attention: Both solution needs to "refresh" the worker file each time you upgrade pdfjs version (or ngx-extended-pdf-viewer when the pdfjs internal version change).

Thank you, and I hope both solutions could help if need it.

**Comment by gallienius** [comment]
Is there an example of how to load pdf.js from ngx-extended-pdf-viewer/assets and use it.

**Comment by stephanrauh** [comment]
@gallienius You don't need to load pdf.js to create thumbnails. Look here: https://pdfviewer.net/extended-pdf-viewer/export

**Comment by cre8** [comment]
@stephanrauh I have the same problem, but I need the pdf.js resource from the assets folder to avoid installing `ngx-extended-pdf` and `pdfjs-dist`

I am not sure how can I reference the pdfjs library, `import * as pdfjs from 'ngx-extended-pdf-viewer/assets/pdf-2.10.552.js'` is the wrong way :(

**Comment by stephanrauh** [comment]
@cre8 Yes, you're probably right. `import *` add the pdfjs code to your `vendor.js` (or something like that). I'm simply loading the pdf.js and viewer.js at runtime by adding a `<script>` tag. I suppose you can do that, too, and by doing so ngx-extended-pdf-viewer recognizes that someone else already has loaded pdf.js:

```typescript
  private loadViewer(): void {
    if (!window['pdfjs-dist/build/pdf']) {
      setTimeout(() => this.loadViewer(), 25); // wait until pdf.js is initialized
    } else {
      const assets = pdfDefaultOptions.assetsFolder;
      script2.src = this.location.normalize(assets + '/viewer-2.10.552.min.js');
      script2.type = 'text/javascript';
      script2.async = true;
      document.getElementsByTagName('head')[0].appendChild(script2);
    }
  }

  private loadPdfJs() {
    if (!window['pdfjs-dist/build/pdf']) {
      const assets = pdfDefaultOptions.assetsFolder;
      const versionSuffix = getVersionSuffix(assets);
      const script = document.createElement('script');
      script.src = this.location.normalize(assets + '/pdf-2.1.552.min.js');
      script.type = 'text/javascript';
      script.async = true;
      document.getElementsByTagName('head')[0].appendChild(script);
    }
    if (!(window as any).webViewerLoad) {
      this.loadViewer();
    }
  }

  public ngOnInit(): void {
    this.loadPdfJs();
  }

```

**Comment by derekkite** [comment]
For someone looking for a more up to date answer to this question.

```
@Injectable()
export class UploadThumbnailsService {
  pdfservice = inject(PDFScriptLoaderService)

  constructor() {
    this.pdfservice.ensurePdfJsHasBeenLoaded(true, false).then(
      () => {
        console.log('pdfjs loaded')
      },
      (err) => {
        console.log('pdfjs not loaded', err)
      }
    )
  }
```

to initialize it.

```
            // @ts-ignore
            const pdfjsLib = window['pdfjsLib']
            pdfjsLib.GlobalWorkerOptions.workerSrc = `/assets/pdf.worker-${this.pdfservice.pdfjsVersion}.mjs`;
            pdfjsLib.getDocument({data: pdfData}).promise.then((pdf: any) => {
```
I needed the worker to create thumbnails.

I was running into a version collision.
